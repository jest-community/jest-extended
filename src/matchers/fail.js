export function fail(_, message) {
  return {
    pass: false,
    message: () => (message ? message : 'fails by .fail() assertion'),
  };
}
