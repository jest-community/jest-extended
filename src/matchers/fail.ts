export function fail(_: unknown, message: string) {
  return {
    pass: false,
    message: () => (message ? message : 'fails by .fail() assertion'),
  };
}
