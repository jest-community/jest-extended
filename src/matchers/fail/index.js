import predicate from './predicate';

const failMessage = message => {
  if (message) return () => message;
  else return () => 'fails by .fail() assertion';
};

export default {
  fail: (expected, message) => ({ pass: predicate(), message: failMessage(message) }),
};
