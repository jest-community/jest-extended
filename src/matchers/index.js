import toBeFalse from './toBeFalse';
import toBeTrue from './toBeTrue';

export default [toBeFalse, toBeTrue].reduce((acc, matcher) => ({ ...acc, ...matcher }), {});
