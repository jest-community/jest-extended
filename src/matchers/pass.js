export function pass(_, message) {
  return {
    pass: true,
    message: () => {
      if (message) return message;
      else return 'passes by .pass() assertion';
    },
  };
}
