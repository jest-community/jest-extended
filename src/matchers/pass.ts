export function pass(_, message) {
  return {
    pass: true,
    message: () => (message ? message : 'passes by .pass() assertion'),
  };
}
