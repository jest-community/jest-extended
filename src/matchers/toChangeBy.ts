/**
 * Use `.toChangeBy` when checking if a value changed by an amount.
 */
export function toChangeBy(mutator: () => unknown | void, checker: () => number | bigint, by: number | bigint = 1) {
  // @ts-expect-error OK to have implicit any for this.utils
  const { printReceived: print, matcherHint: hint } = this.utils;

  const before = checker();
  mutator();
  const after = checker();

  let received;
  if (typeof before === 'bigint' && typeof after === 'bigint') {
    received = after - before;
  } else {
    received = Number(after) - Number(before);
  }

  const passMessage = `
${hint('.not.toChangeBy', 'received', '')}\n\nExpected value to not change by ${by}, received:\n  ${print(received)}
`.trim();
  const failMessage = `
${hint('.toChangeBy', 'received', '')}\n\nExpected value to change by ${by}, received:\n  ${print(received)}
`.trim();

  const pass = received === by;
  return { pass, message: () => (pass ? passMessage : failMessage) };
}
