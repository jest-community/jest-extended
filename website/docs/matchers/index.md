---
sidebar_position: 1
---

# Matchers

- [.pass(message)](/docs/matchers/pass)
- [.fail(message)](/docs/matchers/fail)
- [.toBeEmpty()](/docs/matchers/tobeempty)
- [.toBeOneOf([members])](/docs/matchers/tobeoneof)
- [.toBeNil()](/docs/matchers/tobenil)
- [.toSatisfy(predicate)](/docs/matchers/tosatisfy)

## [Array](/docs/matchers/array)

- [.toBeArray()](/docs/matchers/array/#tobearray)
- [.toBeArrayOfSize()](/docs/matchers/array/#tobearrayofsize)
- [.toIncludeAllMembers([members])](/docs/matchers/array/#toincludeallmembersmembers)
- [.toIncludeAllPartialMembers([members])](/docs/matchers/array/#toincludeallpartialmembersmembers)
- [.toIncludeAnyMembers([members])](/docs/matchers/array/#toincludeanymembersmembers)
- [.toIncludeSameMembers([members])](/docs/matchers/array/#toincludesamemembersmembers)
- [.toPartiallyContain(member)](/docs/matchers/array/#topartiallycontainmember)
- [.toSatisfyAll(predicate)](/docs/matchers/array/#tosatisfyallpredicate)
- [.toSatisfyAny(predicate)](/docs/matchers/array/#tosatisfyanypredicate)
- [.toBeInRange(min, max)](/docs/matchers/array/#tobeinrangemin-max)

## [Boolean](/docs/matchers/boolean)

- [.toBeBoolean()](/docs/matchers/boolean/#tobeboolean)
- [.toBeTrue()](/docs/matchers/boolean/#tobetrue)
- [.toBeFalse()](/docs/matchers/boolean/#tobefalse)

## [Date](/docs/matchers/date)

- [.toBeDate()](/docs/matchers/date/#tobedate)
- [.toBeValidDate()](/docs/matchers/date/#tobevaliddate)
- [.toBeAfter(date)](/docs/matchers/date/#tobeafterdate)
- [.toBeBefore(date)](/docs/matchers/date/#tobebeforedate)
- [.toBeAfterOrEqualTo(date)](/docs/matchers/date/#tobeafterorequaltodate)
- [.toBeBeforeOrEqualTo(date)](/docs/matchers/date/#tobebeforeorequaltodate)
- [.toBeBetween(startDate, endDate)](/docs/matchers/date/#tobebetweenstartdate-enddate)

## [Function](/docs/matchers/function)

- [.toBeFunction()](/docs/matchers/function/#tobefunction)
- [.toThrowWithMessage()](/docs/matchers/function/#tothrowwithmessagetype-message)

## [Mock](/docs/matchers/mock)

- [.toHaveBeenCalledBefore()](/docs/matchers/mock/#tohavebeencalledbefore)
- [.toHaveBeenCalledAfter()](/docs/matchers/mock/#tohavebeencalledafter)
- [.toHaveBeenCalledOnce()](/docs/matchers/mock/#tohavebeencalledonce)
- [.toHaveBeenCalledOnceWith()](/docs/matchers/mock/#tohavebeencalledoncewith)

## [Number](/docs/matchers/number)

- [.toBeNumber()](/docs/matchers/number/#tobenumber)
- [.toBeNaN()](/docs/matchers/number/#tobenan)
- [.toBeFinite()](/docs/matchers/number/#tobefinite)
- [.toBePositive()](/docs/matchers/number/#tobepositive)
- [.toBeNegative()](/docs/matchers/number/#tobenegative)
- [.toBeEven()](/docs/matchers/number/#tobeeven)
- [.toBeOdd()](/docs/matchers/number/#tobeodd)
- [.toBeWithin(start, end)](/docs/matchers/number/#tobewithinstart-end)
- [.toBeInteger()](/docs/matchers/number/#tobeinteger)

## [Object](/docs/matchers/object)

- [.toBeObject()](/docs/matchers/object/#tobeobject)
- [.toBeEmptyObject()](/docs/matchers/object/#tobeemptyobject)
- [.toContainKey(key)](/docs/matchers/object/#tocontainkeykey)
- [.toContainKeys([keys])](/docs/matchers/object/#tocontainkeyskeys)
- [.toContainAllKeys([keys])](/docs/matchers/object/#tocontainallkeyskeys)
- [.toContainAnyKeys([keys])](/docs/matchers/object/#tocontainanykeyskeys)
- [.toContainValue(value)](/docs/matchers/object/#tocontainvaluevalue)
- [.toContainValues([values])](/docs/matchers/object/#tocontainvaluesvalues)
- [.toContainAllValues([values])](/docs/matchers/object/#tocontainallvaluesvalues)
- [.toContainAnyValues([values])](/docs/matchers/object/#tocontainanyvaluesvalues)
- [.toContainEntry([key, value])](/docs/matchers/object/#tocontainentrykey-value)
- [.toContainEntries([[key, value]])](/docs/matchers/object/#tocontainentrieskey-value)
- [.toContainAllEntries([[key, value]])](/docs/matchers/object/#tocontainallentrieskey-value)
- [.toContainAnyEntries([[key, value]])](/docs/matchers/object/#tocontainanyentrieskey-value)
- [.toBeExtensible()](/docs/matchers/object/#tobeextensible)
- [.toBeFrozen()](/docs/matchers/object/#tobefrozen)
- [.toBeSealed()](/docs/matchers/object/#tobesealed)

## [Promise](/docs/matchers/promise)

- [.toResolve()](/docs/matchers/promise/#toresolve)
- [.toReject()](/docs/matchers/promise/#toreject)

## [String](/docs/matchers/string)

- [.toBeString()](/docs/matchers/string/#tobestring)
- [.toBeHexadecimal(string)](/docs/matchers/string/#tobehexadecimal)
- [.toBeDateString(string)](/docs/matchers/string/#tobedatestringstring)
- [.toEqualCaseInsensitive(string)](/docs/matchers/string/#toequalcaseinsensitivestring)
- [.toStartWith(prefix)](/docs/matchers/string/#tostartwithprefix)
- [.toEndWith(suffix)](/docs/matchers/string/#toendwithsuffix)
- [.toInclude(substring)](/docs/matchers/string/#toincludesubstring)
- [.toIncludeRepeated(substring, times)](/docs/matchers/string/#toincluderepeatedsubstring-times)
- [.toIncludeMultiple([substring])](/docs/matchers/string/#toincludemultiplesubstring)
- [.toEqualIgnoringWhitespace(string)](/docs/matchers/string/#toequalignoringwhitespacestring)

## [Symbol](/docs/matchers/symbol)

- [.toBeSymbol()](/docs/matchers/symbol/#tobesymbol)
