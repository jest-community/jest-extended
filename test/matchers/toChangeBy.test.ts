import * as matcher from 'src/matchers/toChangeBy';
import { Counter } from '../fixtures/counter';

expect.extend(matcher);

describe('.toChangeBy', () => {
  test('passes when given a value that the mutator increments', () => {
    let value = 0;
    expect(() => value--).toChangeBy(() => value, -1);
  });

  test('passes when mutating a counter', () => {
    const counter = new Counter();
    expect(counter.increment).toChangeBy(counter.count);
  });

  test('passes when resetting a counter', () => {
    const counter = new Counter();
    counter.increment();
    expect(() => {
      counter.increment();
      counter.increment();
    }).toChangeBy(() => counter.count(), 2);
  });

  test('fails when a given value does not increment', () => {
    const value = 0;
    expect(() => expect(() => value).toChangeBy(() => value)).toThrowErrorMatchingSnapshot();
  });

  test('fails when not mutating a counter', () => {
    const counter = new Counter();
    expect(() => expect(counter.count).toChangeBy(counter.count)).toThrowErrorMatchingSnapshot();
  });

  test('passes when resetting a counter back', () => {
    const counter = new Counter();
    counter.increment();
    counter.increment();
    counter.increment();
    expect(() => counter.reset()).toChangeBy(() => counter.count(), -3);
  });
});

describe('.not.toChangeBy', () => {
  test('passes when given a mutator that does not increment the value', () => {
    const value = 0;
    expect(() => value).not.toChangeBy(() => value, 1);
  });

  test('passes when not mutating a counter', () => {
    const counter = new Counter();
    expect(counter.count).not.toChangeBy(counter.count, 1);
  });

  test('fails when resetting a counter', () => {
    const counter = new Counter();
    expect(() => {
      counter.increment();
      counter.increment();
      counter.increment();
      counter.reset();
    }).not.toChangeBy(() => counter.count(), -3);
  });

  test('fails when mutating a counter', () => {
    const counter = new Counter();
    expect(() => expect(counter.increment).not.toChangeBy(counter.count, 1)).toThrowErrorMatchingSnapshot();
  });

  test('fails when a value expected not to change defies all expectations and changes', () => {
    let value = 0;
    expect(() => expect(() => value++).not.toChangeBy(() => value, 1)).toThrowErrorMatchingSnapshot();
  });
});
