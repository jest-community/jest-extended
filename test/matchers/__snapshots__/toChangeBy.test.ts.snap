// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.not.toChangeBy fails when a value expected not to change defies all expectations and changes 1`] = `
"<dim>expect(</intensity><red>received</color><dim>).not.toChangeBy()</intensity>

Expected value to not change by 1, received:
  <red>1</color>"
`;

exports[`.not.toChangeBy fails when mutating a counter 1`] = `
"<dim>expect(</intensity><red>received</color><dim>).not.toChangeBy()</intensity>

Expected value to not change by 1, received:
  <red>1</color>"
`;

exports[`.toChangeBy fails when a given value does not increment 1`] = `
"<dim>expect(</intensity><red>received</color><dim>).toChangeBy()</intensity>

Expected value to change by 1, received:
  <red>0</color>"
`;

exports[`.toChangeBy fails when not mutating a counter 1`] = `
"<dim>expect(</intensity><red>received</color><dim>).toChangeBy()</intensity>

Expected value to change by 1, received:
  <red>0</color>"
`;
